File: src/v3/runtime/replay.ts
import { DiceRollError } from '../errors.js';
import { MATH_PROFILE, type MathProfile } from '../math.js';
import { DEFAULT_DICE_LIMITS } from './limits.js';

export type SeedInput = number | string;

export type SeedOrigin = 'provided-number' | 'provided-string' | 'crypto';

export type RandomAlgorithm = 'mt19937' | 'xoshiro128ss';

export interface ReplayDescriptor {
  readonly schemaVersion: 2;
  readonly algorithm: RandomAlgorithm;
  readonly algorithmVersion: 1;
  readonly executionVersion: 1;
  readonly mathProfile: MathProfile;
  readonly origin: SeedOrigin;
  readonly seedMaterial: string;
  readonly planFingerprint: string;
}

export interface SeedMaterial {
  /** Internal namespaced representation. Never included in ReplayDescriptor. */
  readonly canonicalSeed: string;
  readonly seedMaterial: string;
  readonly origin: SeedOrigin;
  readonly words: readonly number[];
}

export interface CryptoSource {
  getRandomValues(array: Uint32Array): Uint32Array;
}

export interface ReplayDescriptorOptions {
  readonly algorithm?: RandomAlgorithm;
  readonly planFingerprint?: string;
}

const WORD_COUNT = 4;
const HEX_128_PATTERN = /^[0-9a-f]{32}$/u;
const EMPTY_PLAN_FINGERPRINT = '00000000000000000000000000000000';

function defaultCryptoSource(): CryptoSource | null {
  const cryptoSource = globalThis.crypto;
  return typeof cryptoSource?.getRandomValues === 'function' ? cryptoSource : null;
}

function hashCanonicalSeed(value: string): readonly number[] {
  let hash = 1_779_033_703 ^ value.length;

  for (let index = 0; index < value.length; index += 1) {
    hash = Math.imul(hash ^ value.charCodeAt(index), 3_432_918_353);
    hash = (hash << 13) | (hash >>> 19);
  }

  const words: number[] = [];
  for (let index = 0; index < WORD_COUNT; index += 1) {
    hash = Math.imul(hash ^ (hash >>> 16), 2_246_822_507);
    hash = Math.imul(hash ^ (hash >>> 13), 3_266_489_909);
    hash ^= hash >>> 16;
    words.push(hash >>> 0);
  }

  return Object.freeze(words);
}

function wordsToHex(words: ArrayLike<number>): string {
  let output = '';
  for (let index = 0; index < words.length; index += 1) {
    output += (words[index] ?? 0).toString(16).padStart(8, '0');
  }
  return output;
}

function hexToWords(material: string): readonly number[] {
  const words: number[] = [];
  for (let index = 0; index < WORD_COUNT; index += 1) {
    words.push(Number.parseInt(material.slice(index * 8, (index + 1) * 8), 16));
  }
  return Object.freeze(words);
}

function invalidReplay(message: string): DiceRollError {
  return new DiceRollError(message, { code: 'INVALID_REPLAY' });
}

function isRecord(value: unknown): value is Readonly<Record<string, unknown>> {
  if (typeof value !== 'object' || value === null) {
    return false;
  }
  try {
    return !Array.isArray(value);
  } catch {
    return false;
  }
}

function hasExactReplayKeys(value: Readonly<Record<string, unknown>>): boolean {
  const expected = [
    'algorithm',
    'algorithmVersion',
    'executionVersion',
    'mathProfile',
    'origin',
    'planFingerprint',
    'schemaVersion',
    'seedMaterial',
  ];
  try {
    return Object.keys(value).sort().join(',') === expected.join(',');
  } catch {
    return false;
  }
}

export function isReplayDescriptor(value: unknown): value is ReplayDescriptor {
  if (!isRecord(value) || !hasExactReplayKeys(value)) {
    return false;
  }

  try {
    return value['schemaVersion'] === 2
      && (value['algorithm'] === 'mt19937' || value['algorithm'] === 'xoshiro128ss')
      && value['algorithmVersion'] === 1
      && value['executionVersion'] === 1
      && value['mathProfile'] === MATH_PROFILE
      && (value['origin'] === 'provided-number'
        || value['origin'] === 'provided-string'
        || value['origin'] === 'crypto')
      && typeof value['seedMaterial'] === 'string'
      && HEX_128_PATTERN.test(value['seedMaterial'])
      && typeof value['planFingerprint'] === 'string'
      && HEX_128_PATTERN.test(value['planFingerprint']);
  } catch {
    return false;
  }
}

/** Validates all replay fields and optionally binds it to an expected plan. */
export function validateReplayDescriptor(
  value: unknown,
  expectedPlanFingerprint?: string,
): ReplayDescriptor {
  if (!isRecord(value)) {
    throw invalidReplay('Replay descriptor must be a non-null object');
  }

  try {
    if (value['schemaVersion'] !== 2
      || value['algorithmVersion'] !== 1
      || value['executionVersion'] !== 1
      || value['mathProfile'] !== MATH_PROFILE
      || (value['algorithm'] !== 'mt19937' && value['algorithm'] !== 'xoshiro128ss')) {
      throw new DiceRollError('The replay algorithm or version is not supported', {
        code: 'UNSUPPORTED_REPLAY_VERSION',
      });
    }
  } catch (error: unknown) {
    if (error instanceof DiceRollError) {
      throw error;
    }
    throw invalidReplay('Replay descriptor could not be read');
  }

  if (!isReplayDescriptor(value)) {
    throw invalidReplay('Replay descriptor contains malformed or unexpected fields');
  }
  if (expectedPlanFingerprint !== undefined
    && value.planFingerprint !== expectedPlanFingerprint) {
    throw new DiceRollError('Replay descriptor belongs to a different roll plan', {
      code: 'REPLAY_PLAN_MISMATCH',
      details: {
        expectedPlanFingerprint,
        actualPlanFingerprint: value.planFingerprint,
      },
    });
  }
  return value;
}

export function canonicalizeSeed(
  seed: SeedInput,
  maxSeedLength = DEFAULT_DICE_LIMITS.maxSeedLength,
): string {
  if (!Number.isSafeInteger(maxSeedLength) || maxSeedLength < 1) {
    throw new DiceRollError('maxSeedLength must be a positive safe integer', {
      code: 'INVALID_LIMIT',
      details: { maxSeedLength },
    });
  }

  if (typeof seed === 'number') {
    if (!Number.isFinite(seed)) {
      throw new DiceRollError('Numeric seeds must be finite', {
        code: 'INVALID_SEED',
        details: { seed: String(seed) },
      });
    }

    const value = Object.is(seed, -0) ? '-0' : seed.toString(10);
    return `number:${value}`;
  }

  if (typeof seed !== 'string') {
    throw new DiceRollError('Seeds must be finite numbers or strings', {
      code: 'INVALID_SEED',
    });
  }

  if (seed.length > maxSeedLength) {
    throw new DiceRollError('Text seed exceeds the maximum length', {
      code: 'INVALID_SEED',
      details: { seedLength: seed.length, maxSeedLength },
    });
  }
  return `string:${seed}`;
}

export function createProvidedSeed(
  seed: SeedInput,
  maxSeedLength = DEFAULT_DICE_LIMITS.maxSeedLength,
): SeedMaterial {
  const canonicalSeed = canonicalizeSeed(seed, maxSeedLength);
  const words = hashCanonicalSeed(canonicalSeed);
  return Object.freeze({
    canonicalSeed,
    seedMaterial: wordsToHex(words),
    origin: typeof seed === 'number' ? 'provided-number' : 'provided-string',
    words,
  });
}

export function createAutomaticSeed(
  cryptoSource: CryptoSource | null = defaultCryptoSource(),
): SeedMaterial {
  if (cryptoSource === null) {
    throw new DiceRollError('A cryptographic random source is not available', {
      code: 'RNG_UNAVAILABLE',
    });
  }

  const values = new Uint32Array(WORD_COUNT);
  try {
    cryptoSource.getRandomValues(values);
  } catch {
    throw new DiceRollError('The cryptographic random source failed', {
      code: 'RNG_UNAVAILABLE',
    });
  }
  const words = Object.freeze(Array.from(values));
  const seedMaterial = wordsToHex(words);

  return Object.freeze({
    canonicalSeed: `crypto:${seedMaterial}`,
    seedMaterial,
    origin: 'crypto',
    words,
  });
}

/** Restores the fixed 128-bit seed represented by a public replay descriptor. */
export function createReplaySeed(
  descriptor: unknown,
  expectedPlanFingerprint?: string,
): SeedMaterial {
  const replay = validateReplayDescriptor(descriptor, expectedPlanFingerprint);
  const words = hexToWords(replay.seedMaterial);
  return Object.freeze({
    canonicalSeed: `replay:${replay.seedMaterial}`,
    seedMaterial: replay.seedMaterial,
    origin: replay.origin,
    words,
  });
}

export function createSeedMaterial(
  seed: SeedInput | undefined,
  cryptoSource: CryptoSource | null = defaultCryptoSource(),
  maxSeedLength = DEFAULT_DICE_LIMITS.maxSeedLength,
): SeedMaterial {
  return seed === undefined
    ? createAutomaticSeed(cryptoSource)
    : createProvidedSeed(seed, maxSeedLength);
}

export function createReplayDescriptor(
  seed: SeedMaterial,
  options: ReplayDescriptorOptions = {},
): ReplayDescriptor {
  const algorithm = options.algorithm ?? 'mt19937';
  const planFingerprint = options.planFingerprint ?? EMPTY_PLAN_FINGERPRINT;
  if ((algorithm !== 'mt19937' && algorithm !== 'xoshiro128ss')
    || !HEX_128_PATTERN.test(seed.seedMaterial)
    || !HEX_128_PATTERN.test(planFingerprint)) {
    throw invalidReplay('Replay seed material and plan fingerprint must be 128-bit hex strings');
  }

  return Object.freeze({
    schemaVersion: 2,
    algorithm,
    algorithmVersion: 1,
    executionVersion: 1,
    mathProfile: MATH_PROFILE,
    origin: seed.origin,
    seedMaterial: seed.seedMaterial,
    planFingerprint,
  });
}

File: go/replay.go
package dicecore

import (
	"bytes"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"io"
)

type RandomAlgorithm string

const (
	MT19937              RandomAlgorithm = "mt19937"
	Xoshiro128SS         RandomAlgorithm = "xoshiro128ss"
	EmptyPlanFingerprint                 = "00000000000000000000000000000000"
)

// ReplayDescriptor is the versioned wire contract shared with TypeScript.
// Its fixed seed material excludes the original user seed text.
type ReplayDescriptor struct {
	SchemaVersion    int             `json:"schemaVersion"`
	Algorithm        RandomAlgorithm `json:"algorithm"`
	AlgorithmVersion int             `json:"algorithmVersion"`
	ExecutionVersion int             `json:"executionVersion"`
	MathProfile      string          `json:"mathProfile"`
	Origin           SeedOrigin      `json:"origin"`
	SeedMaterial     string          `json:"seedMaterial"`
	PlanFingerprint  string          `json:"planFingerprint"`
}

type ReplayDescriptorOptions struct {
	Algorithm       RandomAlgorithm
	PlanFingerprint string
}

type ReplayState struct {
	Replay ReplayDescriptor
	Seed   SeedMaterial
}

func validReplayHex(value string) bool {
	if len(value) != 32 {
		return false
	}
	for i := range value {
		if !(value[i] >= '0' && value[i] <= '9') && !(value[i] >= 'a' && value[i] <= 'f') {
			return false
		}
	}
	return true
}

func supportedRandomAlgorithm(algorithm RandomAlgorithm) bool {
	return algorithm == MT19937 || algorithm == Xoshiro128SS
}

func invalidReplayError(message string) error {
	return newDiceError("INVALID_REPLAY", message, "", nil)
}

func CreateReplayDescriptor(seed SeedMaterial, options ReplayDescriptorOptions) (ReplayDescriptor, error) {
	algorithm := options.Algorithm
	if algorithm == "" {
		algorithm = MT19937
	}
	fingerprint := options.PlanFingerprint
	if fingerprint == "" {
		fingerprint = EmptyPlanFingerprint
	}
	if !supportedRandomAlgorithm(algorithm) || !validReplayHex(seed.SeedMaterial) || !validReplayHex(fingerprint) {
		return ReplayDescriptor{}, invalidReplayError("Replay seed material and plan fingerprint must be 128-bit hex strings")
	}
	return ReplayDescriptor{
		SchemaVersion: 2, Algorithm: algorithm, AlgorithmVersion: 1,
		ExecutionVersion: 1, MathProfile: MathProfile, Origin: seed.Origin,
		SeedMaterial: seed.SeedMaterial, PlanFingerprint: fingerprint,
	}, nil
}

// ValidateReplayDescriptor accepts a native descriptor or decoded JSON object.
// JSON decoding also uses this validator, including rejection of extra keys.
// The optional expected fingerprint binds replay to the compiled roll plan.
func ValidateReplayDescriptor(value any, expectedPlanFingerprint ...string) (ReplayDescriptor, error) {
	if len(expectedPlanFingerprint) > 1 {
		return ReplayDescriptor{}, invalidReplayError("Only one expected plan fingerprint may be provided")
	}
	var descriptor ReplayDescriptor
	exact := true
	switch input := value.(type) {
	case ReplayDescriptor:
		descriptor = input
	case *ReplayDescriptor:
		if input == nil {
			return ReplayDescriptor{}, invalidReplayError("Replay descriptor must be a non-null object")
		}
		descriptor = *input
	case map[string]any:
		if input == nil {
			return ReplayDescriptor{}, invalidReplayError("Replay descriptor must be a non-null object")
		}
		// Version validation precedes shape validation, as in TypeScript.
		if !replayVersion(input["schemaVersion"], 2) || !replayVersion(input["algorithmVersion"], 1) ||
			!replayVersion(input["executionVersion"], 1) || input["mathProfile"] != MathProfile ||
			(input["algorithm"] != "mt19937" && input["algorithm"] != "xoshiro128ss") {
			return ReplayDescriptor{}, newDiceError("UNSUPPORTED_REPLAY_VERSION", "The replay algorithm or version is not supported", "", nil)
		}
		algorithm, _ := input["algorithm"].(string)
		origin, originOK := input["origin"].(string)
		seed, seedOK := input["seedMaterial"].(string)
		fingerprint, fingerprintOK := input["planFingerprint"].(string)
		exact = len(input) == 8 && originOK && seedOK && fingerprintOK
		descriptor = ReplayDescriptor{
			SchemaVersion: 2, AlgorithmVersion: 1, ExecutionVersion: 1, MathProfile: MathProfile,
			Algorithm: RandomAlgorithm(algorithm), Origin: SeedOrigin(origin), SeedMaterial: seed, PlanFingerprint: fingerprint,
		}
	default:
		return ReplayDescriptor{}, invalidReplayError("Replay descriptor must be a non-null object")
	}
	if descriptor.SchemaVersion != 2 || descriptor.AlgorithmVersion != 1 || descriptor.ExecutionVersion != 1 ||
		descriptor.MathProfile != MathProfile || !supportedRandomAlgorithm(descriptor.Algorithm) {
		return ReplayDescriptor{}, newDiceError("UNSUPPORTED_REPLAY_VERSION", "The replay algorithm or version is not supported", "", nil)
	}
	if !exact || !validReplayHex(descriptor.SeedMaterial) || !validReplayHex(descriptor.PlanFingerprint) ||
		(descriptor.Origin != SeedProvidedNumber && descriptor.Origin != SeedProvidedString && descriptor.Origin != SeedCrypto) {
		return ReplayDescriptor{}, invalidReplayError("Replay descriptor contains malformed or unexpected fields")
	}
	if len(expectedPlanFingerprint) == 1 && descriptor.PlanFingerprint != expectedPlanFingerprint[0] {
		return ReplayDescriptor{}, newDiceError("REPLAY_PLAN_MISMATCH", "Replay descriptor belongs to a different roll plan", "", map[string]any{
			"expectedPlanFingerprint": expectedPlanFingerprint[0], "actualPlanFingerprint": descriptor.PlanFingerprint,
		})
	}
	return descriptor, nil
}

func replayVersion(value any, expected int) bool {
	switch number := value.(type) {
	case float64:
		return number == float64(expected)
	case int:
		return number == expected
	case json.Number:
		parsed, err := number.Float64()
		return err == nil && parsed == float64(expected)
	default:
		return false
	}
}

func IsReplayDescriptor(value any) bool {
	_, err := ValidateReplayDescriptor(value)
	return err == nil
}

// UnmarshalJSON validates the complete external envelope instead of silently
// dropping unknown fields as encoding/json normally does for a struct.
func (descriptor *ReplayDescriptor) UnmarshalJSON(data []byte) error {
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.UseNumber()
	var value any
	if err := decoder.Decode(&value); err != nil {
		return invalidReplayError("Replay descriptor could not be read")
	}
	var trailing any
	if err := decoder.Decode(&trailing); err != io.EOF {
		return invalidReplayError("Replay descriptor could not be read")
	}
	validated, err := ValidateReplayDescriptor(value)
	if err != nil {
		return err
	}
	*descriptor = validated
	return nil
}

func CreateReplayState(value any, expectedPlanFingerprint ...string) (ReplayState, error) {
	replay, err := ValidateReplayDescriptor(value, expectedPlanFingerprint...)
	if err != nil {
		return ReplayState{}, err
	}
	material, _ := hex.DecodeString(replay.SeedMaterial) // validated lowercase hex
	var words [4]uint32
	for i := range words {
		words[i] = binary.BigEndian.Uint32(material[i*4:])
	}
	return ReplayState{Replay: replay, Seed: SeedMaterial{
		CanonicalSeed: "replay:" + replay.SeedMaterial, SeedMaterial: replay.SeedMaterial,
		Origin: replay.Origin, Words: words,
	}}, nil
}

func CreateReplaySeed(value any, expectedPlanFingerprint ...string) (SeedMaterial, error) {
	state, err := CreateReplayState(value, expectedPlanFingerprint...)
	return state.Seed, err
}

// NewReplayRandom restores a fresh generator from a validated descriptor.
// All state is private to this generator and its supplied execution budget.
func NewReplayRandom(value any, budget RandomCallBudget, expectedPlanFingerprint ...string) (RandomSource, error) {
	state, err := CreateReplayState(value, expectedPlanFingerprint...)
	if err != nil {
		return nil, err
	}
	if state.Replay.Algorithm == MT19937 {
		return NewMersenneTwister19937FromWords(state.Seed.Words[:], budget)
	}
	return NewXoshiro128StarStar(state.Seed.Words[:], budget)
}
